src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +629,128 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
+                    credit >= totalBufferLength - activeBufferLength &&
+                    activeBufferLength > 0)
+                {
+                    // If the buffer has already grown to 32k, does not have room for the next request,
+                    // and is non-empty, flush the current contents to the wire.
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                _outgoingBuffer.EnsureAvailableSpace(credit);
+            }
+            catch
+            {
+                if (acquireWriteLock)
+                {
+                    _writerLock.Release();
+                }
+                throw;
+            }
         }
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
+        // they want to release it.
+        private async ValueTask FinishWriteAsync(bool mustFlush, CancellationToken cancellationToken = default, bool releaseWriteLock = true)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
+                {
+                    if (_inProgressWrite != null)
+                    {
+                        await _inProgressWrite.ConfigureAwait(false);
+                    }
+
+                    _inProgressWrite = FlushOutgoingBytesAsync();
+                    
+                    var tcs = new TaskCompletionSourceWithCancellation<bool>();
+                    
+                    if (_inProgressWrite != await Task.WhenAny(_inProgressWrite, tcs.WaitWithCancellationAsync(cancellationToken)).ConfigureAwait(false))
```

Comment:
I would appreciate some additional scrutiny of this code, since it's the first time I've tried to implement something like this. I based this off of the information [here](https://blogs.msdn.microsoft.com/pfxteam/2012/10/05/how-do-i-cancel-non-cancelable-async-operations/).

Reply:
I don't think we need to wait here at all. We've written the buffer to the kernel's send buffer. We don't care at this point if this completes. We just need to wait for it in StartWriteAsync if it hasn't completed.

Combined with my comment above, I think this means you can (a) remove the cancellation token here and (b) make this function synchronous.

Reply:
To clarify: we haven't necessarily written it to the kernel buffer yet, but that doesn't matter, because we've called WriteAsync on the underlying stream so it will be written at some point. We don't care when the actual write happens. We just need to wait on it in StartWriteAsync, and deal with failures there.

Reply:
> We've written the buffer to the kernel's send buffer.

Is that always going to be true? I know that we have called `SslStream.WriteAsync`, but we haven't awaited it. Will it always copy the data over into the kernel send buffer before returning?

Reply:
Actually, let me revise that. I see what you mean here, and will go ahead and update this section of code.